core/src/main/java/net/sourceforge/jnlp/cache/cache/CacheLRUWrapper.java

```diff
@@ -0,0 +1,540 @@
+ */
+package net.sourceforge.jnlp.cache.cache;
+
+import net.adoptopenjdk.icedteaweb.StringUtils;
+import net.adoptopenjdk.icedteaweb.io.FileUtils;
+import net.adoptopenjdk.icedteaweb.io.IOUtils;
+import net.adoptopenjdk.icedteaweb.jnlp.version.VersionId;
+import net.adoptopenjdk.icedteaweb.jnlp.version.VersionString;
+import net.adoptopenjdk.icedteaweb.logging.Logger;
+import net.adoptopenjdk.icedteaweb.logging.LoggerFactory;
+import net.adoptopenjdk.icedteaweb.os.OsUtil;
+import net.sourceforge.jnlp.config.ConfigurationConstants;
+import net.sourceforge.jnlp.config.InfrastructureFileDescriptor;
+import net.sourceforge.jnlp.config.PathsAndFiles;
+import net.sourceforge.jnlp.runtime.JNLPRuntime;
+import net.sourceforge.jnlp.util.WindowsShortcutManager;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URL;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static net.adoptopenjdk.icedteaweb.CollectionUtils.isEmpty;
+
+/**
+ * This class helps maintain the ordering of most recently use cache items across
+ * multiple jvm instances. The LRU information is stored as a properties file in the
+ * root of the file cache directory. The property key is a combination of a timestamp
+ * and the cache folder id. The property value is the path to the cached item.
+ */
+class CacheLRUWrapper {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CacheLRUWrapper.class);
+
+    static CacheLRUWrapper getInstance() {
+        return CacheLRUWrapperHolder.INSTANCE;
+    }
+
+    private final LeastRecentlyUsedCacheIndexHolder cacheIndex;
+    private final InfrastructureFileDescriptor rootCacheDir;
+
+    /**
+     * @param recentlyUsed file to be used as recently_used file
+     * @param cacheDir     dir with cache
+     */
+    private CacheLRUWrapper(final InfrastructureFileDescriptor recentlyUsed, final InfrastructureFileDescriptor cacheDir) {
+        this.cacheIndex = new LeastRecentlyUsedCacheIndexHolder(recentlyUsed);
+        this.rootCacheDir = cacheDir;
+    }
+
+    File getOrCreateCacheFile(URL resourceHref, VersionId version) {
+        final LeastRecentlyUsedCacheEntry entry = cacheIndex.getSynchronized(idx ->
+                getOrCreateCacheEntry(idx, resourceHref, version)
+        );
+        return getCacheFile(entry);
+    }
+
+    private LeastRecentlyUsedCacheEntry getOrCreateCacheEntry(LeastRecentlyUsedCacheIndex idx, URL resourceHref, VersionId version) {
+        return idx.findAndMarkAsAccessed(resourceHref, version)
```

Comment:
null check for idx

Reply:
internal use with non null arguments only